doc/design/proposal-moving-hooks-on-thread.md

```diff
@@ -0,0 +1,160 @@
+
+## Design
+
+This proposal builds on the [Universal, synchronous and in-thread loader hooks proposal](https://github.com/nodejs/loaders/pull/198). First we would implement the new `registerHooks` API from that proposal, which creates sync `resolve`, `load` and other hooks. That should cover that proposal’s primary goal of providing new APIs for monkey-patchers to migrate to, allowing us to sunset the need to support monkey-patching the CommonJS loader.
+
+Once that is achieved, there is one large gap left to fill before we can remove the off-thread hooks: provide a way to perform async work, now that the hooks are all sync.
+
+### `startGraph` hook (module graph entry point only)
+
+We would create a new _async_ hook that runs for the main entry point, any worker entry point, and any dynamic `import()`: in other words, before any new module graph is created. This hook would delay any further hooks for that module graph operation, until the `startGraph` hook promise has resolved successfully. This provides an opportunity to do async work that the later sync hooks could have access to, for example to make network calls in a non-blocking way. One of the use cases for this hook is to support network imports, such as the [Import from HTTPS example](https://nodejs.org/api/module.html#import-from-https) in our docs. See [`nodejs/node` #43245](https://github.com/nodejs/node/pull/43245).
+
+Because this hook might need to resolve and load the entire module graph to do its work, such as to resolve and fetch HTTPS URLs, it will be provided with the `resolve` and `load` hooks in its `context`. It would also receive `context.entry` to distinguish main/worker entry points from dynamic `import()`. Its signature would be something like:
+
+```js
+/**
+ * @typedef {{
+ *   parentURL?: string,
+ *   entry: boolean,
+ *   conditions?: string[],
+ *   importAttributes?: Record<string, string>,
+ *   resolve: Function,
+ *   load: Function,
+ * }} ModuleStartGraphContext
+ */
+/**
+ * @typedef {Promise<{
+ *   shortCircuit?: boolean
+ * }>} ModuleStartGraphResult
+ */
+/**
+ * @param {string} specifier
+ * @param {ModuleStartGraphContext} context
+ * @param {(specifier: string, context: ModuleStartGraphContext) => ModuleStartGraphResult} nextStartGraph
+ * @returns {ModuleStartGraphResult}
+ */
+async function startGraph(specifier, context, nextStartGraph) {}
+```
+
+## Examples
+
+### Import from HTTPS
+
+the [Import from HTTPS example](https://nodejs.org/api/module.html#import-from-https) in our docs could use `startGraph` to handle the async work:
+
+```js
+import { parseAllImportSpecifiersFromSource } from 'some-library';
+
+/** @type {Map<URL['href'], string | Buffer} */
+const httpImportsCache = new Map();
+
+export async function startGraph(specifier, context, nextStartGraph) {
```

Comment:
What does `nextStartGraph` provide that a simple `done()` callback does not?

Reply:
1. It follows the pattern of the other hooks.
2. If multiple sets of hooks define `startGraph`, you can return `{ shortCircuit: true }` from yours to prevent the others from running.